Set device and gateway location on map

pgalic96 opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #1380

#### Changes
<!-- What are the changes made in this pull request? -->

- Add `react-leaflet` which allows us to set `onClick` events and propagate values correctly without doing vanilla javascript magic to update markers
- Add map to location form and propagate values from map to fields and reverse

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->
- There is one issue, since the map is re-rendered every time a marker is set on the map, the zoom value gets reset. Suggestions how to avoid this re-render while still reset the markers are welcome cc @kschiffer @bafonins 

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

bafonins commented on pkg/webui/components/map/index.js:
```diff
@@ -34,75 +33,62 @@ Leaflet.Icon.Default.mergeOptions({
+      <Map
+        className={classnames(style.map, { [style.widget]: widget })}
+        center={position}
+        zoom={7}
```
can you try saving the zoom value as well? i think it might solve the issue with resetting zoom value after setting a marker

pgalic96 replied:
fixed by storing the local state of the variable. had to move it under `{...leafletConfig}` in component to avoid overwriting in case the zoom value is set in a parent component.

bafonins commented on pkg/webui/console/components/location-form/index.js:
```diff
@@ -129,6 +132,23 @@ class LocationForm extends Component {
+  handleClick(event) {
+    this.setState({ latitude: event.latlng.lat, longitude: event.latlng.lng })
+    this.form.current.setFieldValue('latitude', event.latlng.lat)
+    this.form.current.setFieldValue('longitude', event.latlng.lng)
```
Also, you can use `setValues` here

bafonins commented on pkg/webui/components/map/index.js:
```diff
@@ -34,75 +33,62 @@ Leaflet.Icon.Default.mergeOptions({
+      markers.length >= 1
+        ? [markers[0].position.latitude, markers[0].position.longitude]
+        : [52.7, 4.9]
+    const map = (
```
You dont really need a local variable for this. why not wrapping it directly in the div container?

bafonins commented:
Could you update `leaflet` to 1.6.0 ?

kschiffer commented on pkg/webui/components/map/index.js:
```diff
@@ -34,75 +33,62 @@ Leaflet.Icon.Default.mergeOptions({
+        {...leafletConfig}
+      >
+        <TileLayer
+          url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
```
This should also be a constant.

kschiffer commented on pkg/webui/console/components/location-form/index.js:
```diff
@@ -155,10 +175,22 @@ class LocationForm extends Component {
 
     const entryExists = hasLocationSet(initialValues)
 
+    const marker =
+      latitude && longitude
+        ? [
+            {
+              position: {
+                latitude: Number(latitude),
+                longitude: Number(longitude),
+              },
+            },
+          ]
+        : undefined
+
```
I think we should look into avoiding multiline ternaries.
```suggestion
    let marker = undefined
    if (latitude && longitude) {
      marker =  [{
        position: {
          latitude: Number(latitude),
          longitude: Number(longitude),
        },
      }]
```

pgalic96 replied:
this needs to be fixed in prettier config, otherwise travis will complain

kschiffer replied:
The styling is arbitrary, I was talking about replacing multiline ternaries with if statements.

pgalic96 replied:
aah, sorry noticed the wrong thing

pgalic96 commented:
@kschiffer 
Fixed stuff according to your feedback.

> Clicking on the map now leads to a weird slide. Seems like an issue with the dimensions of the map vs the dimensions of the container but 🤷‍♂️)

This is due to recentering the map based on the marker position. I fixed this, but as you can see, this is not so convenient when setting the location with latitude and longitude fields, as map does not automatically recenter to that field. I can try fixing this, but I think we should not overengineer this logic, for the sake of a very small UX improvement. Your decision, I don't think that this sliding is annoying tbh.

bafonins commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +35,111 @@ Leaflet.Icon.Default.mergeOptions({
-      Leaflet.marker([marker.position.latitude, marker.position.longitude]).addTo(this.map),
-    )
+  getCurrentLocation() {
+    if ('geolocation' in navigator) {
```
Can you create a react hook to access data from the geolocation api?

bafonins commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +35,111 @@ Leaflet.Icon.Default.mergeOptions({
+  getCurrentLocation() {
+    if ('geolocation' in navigator) {
+      navigator.geolocation.getCurrentPosition(function(position) {
+        return [position.coords.latitude, position.coords.longitude]
```
you can also read the altitude value

kschiffer commented:
> @kschiffer
> Fixed stuff according to your feedback.
> 
> > Clicking on the map now leads to a weird slide. Seems like an issue with the dimensions of the map vs the dimensions of the container but 🤷‍♂️)
> 
> This is due to recentering the map based on the marker position. I fixed this, but as you can see, this is not so convenient when setting the location with latitude and longitude fields, as map does not automatically recenter to that field. I can try fixing this, but I think we should not overengineer this logic, for the sake of a very small UX improvement. Your decision, I don't think that this sliding is annoying tbh. My suggestion is to revert this change and let it recenter each time the marker is set.

You should be able to hook into the `onBlur` event of the inputs to trigger the update in the map.

I get asked now for permission to use my location, but the map does not update after that and stays in Amsterdam. See my code comment about that. 
Also, the cursor in the map should be the default one. Otherwise, it's hard to target the precise position.

pgalic96 commented:
so, I resolved most of the issues that we discussed:
- introduced new prop in location form `mapCenter` that gets updated only when the field input changes. this is connected to `mapCenter` state variable in the map itself via `getDerivedStateFromProps` and its used for centering the map. So, the map gets recentered only when the field change because the `mapCenter` doesn't get updated on map click.

- added new prop `clickable` to map which allows us to avoid `onClick` handler if the map is not clickable, and also changed the cursor to pointer (maybe a bit more user friendly than the default one?)

- regarding the geolocation, the location is updated asynchronously by updating the `mapCenter` state variable. I added the overlay component, but the leaflet css file has hardcoded `z-values` which conflict hard with overlay z-values so this component is not ideal for this case. any ideas how to approach this issue?

cc @kschiffer @bafonins

kschiffer commented on pkg/webui/components/map/map.styl:
```diff
@@ -20,6 +20,9 @@
   height: 25rem
 
+.click
+  cursor: pointer !important
```
Any way to get this to work without `!important`? (Though don't invest more than 5 minutes in that)

kschiffer commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +36,143 @@ Leaflet.Icon.Default.mergeOptions({
+      zoom: PropTypes.number,
+    }),
+    mapCenter: PropTypes.arrayOf(PropTypes.number),
     // Markers is an array of objects containing a specific properties
```
```suggestion
    // Markers is an array of objects containing a specific properties.
```

kschiffer commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +36,143 @@ Leaflet.Icon.Default.mergeOptions({
 })
 
-export default class Map extends React.Component {
+const defaultLocation = [52.3, 4.8]
```
Let's use a neutral default location `[0,0]`

kschiffer commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +36,143 @@ Leaflet.Icon.Default.mergeOptions({
+      return
+    }
+    const map = this.refs.map.leafletElement
+    this.setState({ mapCenter: map.getCenter() })
```
I think the map component itself should stay neutral. If we make too many assumptions about the map behavior we will have a hard time using the map for other purposes too. You can already see that as your implementation now needs the `clickable` prop. The map should be first and foremost concerned with rendering a map based on its props.

So instead of keeping track of the map center in the state in the map component itself, that should be done by a parent component (in this case the `<LocationForm />`)

kschiffer commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +36,143 @@ Leaflet.Icon.Default.mergeOptions({
-    })
+    this.state = {
+      zoomLevel: props.leafletConfig.zoom,
+      mapCenter: defaultLocation,
```
What about the location from the marker? Currently the map widget on the overview just stays at the default location, instead of centering on the (first) marker.

pgalic96 commented:
@kschiffer Moved the whole centering logic to location form. Indeed this is a much leaner approach.

kschiffer commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +35,110 @@ Leaflet.Icon.Default.mergeOptions({
-    id: PropTypes.string.isRequired,
+    className: PropTypes.string,
+    clickable: PropTypes.bool,
     // LeafletConfig is an object which can contain any number of properties defined by the leaflet plugin and is used to overwrite the default configuration of leaflet.
```
```suggestion
    // `LeafletConfig` is an object which can contain any number of properties defined by the leaflet plugin and is used to overwrite the default configuration of leaflet.
```

kschiffer commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +35,110 @@ Leaflet.Icon.Default.mergeOptions({
+    mapRef: PropTypes.string,
+    // Markers is an array of objects containing a specific properties.
     markers: PropTypes.arrayOf(
       // Position is a object containing two properties latitude and longitude which are both numbers.
```
```suggestion
      // `position` is an object containing two properties latitude and longitude which are both numbers.
```

kschiffer commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +35,110 @@ Leaflet.Icon.Default.mergeOptions({
+    }),
+    mapCenter: PropTypes.arrayOf(PropTypes.number),
+    mapRef: PropTypes.string,
+    // Markers is an array of objects containing a specific properties.
```
```suggestion
    // `markers` is an array of objects containing a specific properties.
```

kschiffer commented on pkg/webui/components/map/index.js:
```diff
@@ -34,74 +35,110 @@ Leaflet.Icon.Default.mergeOptions({
-    ).isRequired,
+    ),
+    onClick: PropTypes.func,
     // Widget is a boolean used to add a class name to the map container div for styling.
```
```suggestion
    // `widget` is a boolean used to add a class name to the map container div for styling.
```

kschiffer commented on pkg/webui/components/map/widget/index.js:
```diff
@@ -42,10 +42,19 @@ export default class MapWidget extends React.Component {
+      zoom: 7,
+      minZoom: 1,
     }
+    const mapCenter = [markers[0].position.latitude, markers[0].position.longitude]
```
```suggestion
    const mapCenter = Boolean(markers) && markers.length !== 0 ? [markers.length && markers[0].position.latitude, markers[0].position.longitude] : defaultLocation
```
This will crash the console if there is no location set. Please fix and double check that there is no issues with different location configurations.

kschiffer commented on pkg/webui/components/overlay/index.js:
```diff
@@ -22,15 +22,15 @@ import Message from '../../lib/components/message'
     {visible && loading && (
-      <Spinner center>
+      <Spinner center className={spinnerClassName}>
         <Message content={sharedMessages.fetching} />
```
The message of the spinner should be passable (while defaulting to `sharedMessages.fetching`). In case of the location map it should say `Loading location…`
